new blog post: Hello, HPy (#3)
hpyproject/hpyproject.org: posts/2021/03/hello-hpy.md

<!--
.. title: Hello, HPy
.. slug: hello-hpy
.. date: 2021-03-29 10:00:00 UTC
.. author: Antonio Cuni
.. tags: 
.. category: 
.. link: 
.. description: 
.. type: text
-->

Welcome to the shiny new HPy blog!

HPy has been around for a while now. The initial discussion started during
EuroPython 2019, at the good old times when we could still go to conferences
and have real-life meetings. Since then, HPy progressed **a lot** from the
point of view of the actual code, but we have been a bit too silent
w.r.t. communicating what we are doing to the external world and to the broader
Python community. Hopefully, now that this blog is online we will do a better
job at periodically communicating the status of HPy, so make sure to
[subscribe to the RSS feed](/rss.xml).

<!--TEASER_END-->

What is HPy?
============

Quoting the frontpage of our website:

> HPy provides a new API for extending Python in C. In other words, you use
> `#include <hpy.h>` instead of `#include <Python.h>`.

The official [Python/C API](https://docs.python.org/3/c-api/index.html>) is

> specific to the current implementation of CPython: it exposes a lot of

Reviewer: ```suggestion
specific to the current implementation of CPython: it exposes many
```

internal details which makes it hard:

  - to implement it for other Python implementations (e.g. PyPy, GraalPython,
    Jython, IronPython, etc.)

  - to experiment with new things inside CPython itself: e.g. using a GC
    instead of refcounting, or to remove the GIL.

> 

Reviewer: - to correctly check things like refcount handling: the external API gets mixed in with implementation details that should be hidded.

> 

Author: ```suggestion
  - to correctly check things like refcount handling: the external API gets mixed in with implementation details that should be hidded.
```

Over the years, it has become evident that
[emulating the Python/C API in an efficient way is challenging](https://www.pypy.org/posts/2018/09/inside-cpyext-why-emulating-cpython-c-8083064623681286567.html),

> if not impossible. The main goal of HPy is provide a C API which is possible
> to implement in an efficient way on a number of very diverse
> implementations.

Reviewer: ```suggestion
if not impossible. The main goal of HPy is expose a C API which is possible
to implement in an efficient way on a number of very diverse Python
implementations.
```

> There are several advantages to write your C extension in HPy:

Reviewer 2: ```suggestion
There are several advantages to writing your C extension in HPy:
```

  - it runs much faster on PyPy, and at native speed on CPython

  - it is possible to compile a single binary which runs unmodified on all

>     supported Python implementations and versions

Reviewer: Won't this require coordination with the packaging people to create a tag for universal HPy wheels?

Author: we can *compile* a binary even now. To *distribute* the binary yes, it needs cooperation with pypi/wheel/pip/whatever

    invalid usage of objects after they have been deleted. It is possible to
    turn the "debug mode" on at startup time, without needing to recompile
    Python or the extension itself

See also the official docs for a more in-depth
[overview](https://docs.hpyproject.org/en/latest/overview.html#hpy-overview).

Show me an example
===================

This is a "normal" Python/C extension:

{{% listing 2021/03/hello-hpy/hello_old.c c %}}

We can compile it with a simple `setup.py`:

```python
from setuptools import setup, Extension
setup(
    name="hello",
    ext_modules = [
        Extension('hello_old', ['hello_old.c']),
    ],
)
```

```pycon
$ python setup.py build_ext --inplace
...
$ python 
>>> import hello_old
>>> hello_old.add(10, 20)
30
```

Now, let's start to uncover the HPy equivalent, piece by piece:

```c
#include <hpy.h>

HPyDef_METH(add, "add", add_impl, HPyFunc_VARARGS,
            .doc = "add two integers");

static HPy add_impl(HPyContext ctx, HPy self, HPy *args, HPy_ssize_t nargs) {
    long a, b;
    if (!HPyArg_Parse(ctx, NULL, args, nargs, "ll", &a, &b))
        return HPy_NULL;
    return HPyLong_FromLong(ctx, a+b);
}
```

> There are a bunch of things which immediately pop to the eyes:

Reviewer: ```suggestion
There are a bunch of things which are different from the usual C-extension module:
```

- the former `PyObject *` is now `HPy`, which we call "a handle". Handles are
  similar to `PyObject *`, but are completely opaque: for more information,
  see the
  [official docs](https://docs.hpyproject.org/en/latest/api.html#handles).

- There is an additional parameter, `HPyContext ctx`. One of the problems of
  the old API is that often it implicitly relies on the existence of a

>   per-thread or per-subinterpreter local state. `HPyContext` make it explicit,
>   makes the whole API more regular and make it possible to develop new

Reviewer 2: ```suggestion
  per-thread or per-subinterpreter local state. `HPyContext` makes this state explicit. This
  makes the whole API more regular and makes it possible to develop new
```

  a Python tuple. This means that it is possible to call the function without
  having to allocate a Python tuple, and for example the PyPy implementation
  of HPy takes advantage of that. This is very similar to CPython's
  [VectorCall protocol](https://www.python.org/dev/peps/pep-0590/).

!!! Note

    In this post, we are using a slightly old version of HPy. If you try with
    a newer version you should use `HPyContext *ctx` instead of `HPyContext ctx`.
    See also [Issue #150](https://github.com/hpyproject/hpy/issues/150)
    and [PR #182](https://github.com/hpyproject/hpy/pull/182).

Let's continue our tour of `hello_new.c`:

```c
static HPyDef *hello_defines[] = {
    &add,
    NULL
};

static HPyModuleDef moduledef = {
    HPyModuleDef_HEAD_INIT,
    .m_name = "hello_old",
    .m_doc = "hello example using the new HPy API",
    .m_size = -1,
    .defines = hello_defines,
};

HPy_MODINIT(hello_new)
static HPy init_hello_new_impl(HPyContext ctx) {
    return HPyModule_Create(ctx, &moduledef);
}
```

This is pretty similar to the old code. The biggest change is that instead of
declaring an array of `PyMethodDef`, we create an array of `HPyDef` as
discussed above.

> Finally, we need to modify `setup.py`: compiling an HPy extension is as easy

Reviewer: ```suggestion
Finally, we need to modify `setup.py`. Compiling an HPy extension is as easy
```

70
>>> hello_new.__file__
'/.../hello-hpy/hello_new.hpy.so'
```

!!! note

    At the moment of writing, because of
    [Issue #191](https://github.com/hpyproject/hpy/issues/191) if you try to
    print the repr of `hello_new`, you see something like this (note the `.py`
    extension):

        >>> hello_new
        <module 'hello_new' from '/.../hello-hpy/hello_new.py'>

Note that on its own, CPython does **not** know how to import `.hpy.so`
files. The magic is done by the `hello_new.py`, which is automatically
generated by `setup.py`:

```python
$ cat hello_new.py
[...]
def __bootstrap__():
    [...]
    from hpy.universal import load_from_spec
    ext_filepath = pkg_resources.resource_filename(__name__, 'hello_new.hpy.so')
    m = load_from_spec(Spec('hello_new', ext_filepath))
    [...]
    sys.modules[__name__] = m

__bootstrap__()
```

Finally, we can try to import our shiny new universal binary on PyPy:

```shell
$ /path/to/pypy-c-jit-101860-a2f7c80062e8-linux64/bin/pypy
>>>> import hello_new
>>>> hello_new.add(10, 20)
30
>>>> hello_new.__file__
'/.../hello-hpy/hello_new.hpy.so'
>>>>
```

That's all you need to get started with HPy. What we presented today is just

> the basic of course: in the next posts we will dig more into the technical

Reviewer 2: ```suggestion
the basics, of course: in the next posts we will dig more into the technical
```

<!--
.. title: Hello, HPy
.. slug: hello-hpy
.. date: 2021-03-29 10:00:00 UTC
.. author: antocuni
.. tags: 
.. category: 
.. link: 
.. description: 
.. type: text
-->

Welcome to the shiny new HPy blog!

HPy has been around for a while now. The initial discussion started during

> EuroPython 2019, at the good old times when we could still go to conferences

Reviewer 3: ```suggestion
EuroPython 2019, in the good old times when we could still go to conferences
```

and have real-life meetings. Since then, HPy progressed **a lot** from the
point of view of the actual code, but we have been a bit too silent
w.r.t. communicating what we are doing to the external world and to the broader
Python community. Hopefully, now that this blog is online we will do a better
job at periodically communicating the status of HPy, so make sure to
[subscribe to the RSS feed](/rss.xml).

<!--TEASER_END-->

What is HPy?
============

Quoting the frontpage of our website:

> HPy provides a new API for extending Python in C. In other words, you use
> `#include <hpy.h>` instead of `#include <Python.h>`.

The official [Python/C API](https://docs.python.org/3/c-api/index.html>) is
specific to the current implementation of CPython: it exposes many
internal details which makes it hard:

  - to implement it for other Python implementations (e.g. PyPy, GraalPython,
    Jython, IronPython, etc.)

  - to experiment with new things inside CPython itself: e.g. using a GC
    instead of refcounting, or to remove the GIL.

>   - to correctly check things like refcount handling: the external API gets mixed in with implementation details that should be hidded.

Reviewer 3: ```suggestion
  - to correctly check things like refcount handling: the external API gets mixed in with implementation details that should be hidden.
```
